### PR TITLE
Add - snippet edit form

### PR DESCRIPTION
**Pull request by alex-bu-93**

### CODE REVIEW

File: apps/angular-thirty-seconds/src/app/shared/services/github.service.ts [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+      .patch<any>(
+        `${this.apiGithubUrl}/repos/${owner}/${repoName}/issues/${issueNumber}`,
+        {labels: ['snippet']},
+        this.options
+      )
+      .pipe(catchError(() => throwError(new Error(`Can't get issues list`))));
+  }
+
+  getPullFileByPullNumber(owner: string, repoName: string, pullNumber: number): Observable<any> {
+    return this.http
```
Comment (kirjs) [inline apps/angular-thirty-seconds/src/app/shared/services/github.service.ts]: couldn't you use something like:
https://github.com/github-tools/github ?

File: apps/angular-thirty-seconds/src/app/create-snippet/create-snippet.component.html [hunk 2]
```diff
@@ -1,9 +1,7 @@
-<div class="container">
-  <h1>New Snippet
-    <button (click)="onSubmit()"
-            mat-button
-            class="btn-submit">
-      Generate Snippet
+<div *ngIf="!isLoading else loadingTmp" class="container">
+  <h1>{{isEdit ? '' : 'New'}} Snippet
+    <button (click)="onSubmit()" mat-button class="btn-submit">
```
Comment (NothingEverHappens) [inline apps/angular-thirty-seconds/src/app/create-snippet/create-snippet.component.html]: onSubmit -> whatever it actually does?

File: apps/angular-thirty-seconds/src/app/create-snippet/create-snippet.component.html [hunk 3]
```diff
@@ -1,9 +1,7 @@
...
+      {{isEdit ? 'Edit' : 'Create'}} Snippet
```
Comment (NothingEverHappens) [inline apps/angular-thirty-seconds/src/app/create-snippet/create-snippet.component.html]: isEdit -> isEditing 
Есть Редактура -> Редактируется

File: apps/angular-thirty-seconds/src/app/create-snippet/create-snippet.component.ts [hunk 4]
```diff
@@ -1,36 +1,63 @@
...
+import { ValidationsService } from '../shared/services/validations.service';
+import { REPO_NAME, REPO_OWNER } from '../shared/constants/repo-info';
 
 
-function validatorMaxLines(lines: number) {
-  return (control: AbstractControl) => {
-    return control.value.split('\n').length > lines ? {linesError: `This field shouldn't have more than ${lines} lines`} : null;
-  };
+// @ts-ignore
+window.Buffer = {
```
Comment (NothingEverHappens) [inline apps/angular-thirty-seconds/src/app/create-snippet/create-snippet.component.ts]: Add a comment for why this is here

File: apps/angular-thirty-seconds/src/app/create-snippet/create-snippet.component.ts [hunk 5]
```diff
@@ -55,22 +82,113 @@ export class CreateSnippetComponent {
...
   constructor(
     private fb: FormBuilder,
+    private cd: ChangeDetectorRef,
+    private activatedRoute: ActivatedRoute,
+    private snippetService: SnippetService,
+    private validationsService: ValidationsService,
+    private githubService: GitHubService,
     public dialog: MatDialog
   ) {
+    const pullNumber = this.activatedRoute.snapshot.params['itemId'];
```
Comment (NothingEverHappens) [inline apps/angular-thirty-seconds/src/app/create-snippet/create-snippet.component.ts]: itemId or pullNumber? Chose one, use everywhere

File: apps/angular-thirty-seconds/src/app/create-snippet/create-snippet.component.ts [hunk 6]
```diff
@@ -54,21 +80,112 @@ export class CreateSnippetComponent {
...
+    this.destroy.next(null);
+    this.destroy.complete();
+  }
+
+  getPullFileByPullNumber(pullNumber) {
+    this.isLoading = true;
+    this.githubService.getPullFileByPullNumber('nycJSorg', '30-seconds-of-angular', pullNumber)
+      .pipe(
+        takeUntil(this.destroy),
+        switchMap(res => {
```
[the remark this answers is not in the document]
Comment (NothingEverHappens) [reply apps/angular-thirty-seconds/src/app/create-snippet/create-snippet.component.ts]: ☝️

File: apps/angular-thirty-seconds/src/app/create-snippet/create-snippet.component.ts [hunk 7]
```diff
@@ -55,22 +82,113 @@ export class CreateSnippetComponent {
...
+            const file = res[0];
+            this.snippetFileInfo = {
+              sha: file['sha'],
+              fileName: file['filename']
+            };
+            return this.githubService.getSnippetBody(file['contents_url'])
+              .pipe(
+                map(res => {
```
Comment (NothingEverHappens) [inline apps/angular-thirty-seconds/src/app/create-snippet/create-snippet.component.ts]: how does this work?
Pretty sure res overshadows res

File: apps/angular-thirty-seconds/src/app/create-snippet/create-snippet.component.ts [hunk 8]
```diff
@@ -54,21 +80,112 @@ export class CreateSnippetComponent {
...
+                  const body = atob(res.content);
+                  return {...res[0], body};
+                }));
+          }
+        ),
+        switchMap(res => {
+            this.snippetFileInfo['snippet'] = mdTextToJson(res['body']);
```
[the remark this answers is not in the document]
Comment (NothingEverHappens) [reply apps/angular-thirty-seconds/src/app/create-snippet/create-snippet.component.ts]: ☝️

File: apps/angular-thirty-seconds/src/app/shared/services/validations.service.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
+import { Injectable } from '@angular/core';
+import { AbstractControl, FormArray, FormControl, FormGroup } from '@angular/forms';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ValidationsService {
+
+  markFormControlsAsTouched(formGroup: FormGroup | FormArray): void {
```
Comment (NothingEverHappens) [inline apps/angular-thirty-seconds/src/app/shared/services/validations.service.ts]: None of this should be in a service, they all are static functions

File: apps/angular-thirty-seconds/src/app/pull-requests-list/pull-requests-list.component.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  constructor(
+    private router: Router,
+    private githubService: GitHubService
+  ) {
+    this.getPullsList();
+  }
+
+  getPullsList() {
+    this.githubService.getPullsList(this.REPO_OWNER, this.REPO_NAME)
+      .subscribe(res => this.pullsList = res.filter(x => x['labels'].length && x['labels'].map(y => y['name']).indexOf('snippet') > -1))
```
Comment (NothingEverHappens) [inline apps/angular-thirty-seconds/src/app/pull-requests-list/pull-requests-list.component.ts]: Can you use `| async` instead of subscribe

Comment (NothingEverHappens) [reply apps/angular-thirty-seconds/src/app/pull-requests-list/pull-requests-list.component.ts]: this.githubService.getPullsList(this.REPO_OWNER, this.REPO_NAME).pipe(
map(res => res.fliter(....)))

Comment (NothingEverHappens) [reply apps/angular-thirty-seconds/src/app/pull-requests-list/pull-requests-list.component.ts]: map(res => res.fliter(....)) -> move to service

File: apps/angular-thirty-seconds/src/app/shared/services/github.service.ts [hunk 11]
```diff
@@ -0,0 +1,191 @@
+import { Injectable } from '@angular/core';
+import { HttpClient } from '@angular/common/http';
+import { Observable } from 'rxjs/internal/Observable';
+import { catchError } from 'rxjs/operators';
+import { throwError } from 'rxjs/internal/observable/throwError';
+import { Branch, CommitInfo, CreatePullRequest, Repo, User } from '../interfaces';
+import { MatSnackBar } from '@angular/material';
+
+type Commit = any;
```
Comment (NothingEverHappens) [inline apps/angular-thirty-seconds/src/app/shared/services/github.service.ts]: Create another file lol

File: apps/angular-thirty-seconds/src/app/shared/services/github.service.ts [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+      .pipe(catchError(() => this.showErrorSnakeBar(`Can't fork 30 secs repo`)));
+  }
+
+  getMasterBranch(repo: Repo): Observable<Branch> {
+    requires(repo, 'Repository is required');
+
+    const requestUrl = `${this.apiGithubUrl}/repos/${repo.full_name}/git/refs/heads/master`;
+    return this.http
+      .get<Branch>(requestUrl, this.options)
+      .pipe(catchError(() => this.showErrorSnakeBar(`Can't fetch master branch of ${repo.full_name}`)));
```
Comment (NothingEverHappens) [inline apps/angular-thirty-seconds/src/app/shared/services/github.service.ts]: Optionally: Create an operator (or add a todo)

File: apps/angular-thirty-seconds/src/app/shared/services/validations.service.ts [hunk 13]
```diff
@@ -0,0 +1,20 @@
+import { Injectable } from '@angular/core';
+import { AbstractControl } from '@angular/forms';
+
+@Injectable({
```
Comment (NothingEverHappens) [inline apps/angular-thirty-seconds/src/app/shared/services/validations.service.ts]: drop injectable

File: apps/angular-thirty-seconds/src/app/create-snippet/create-snippet.component.ts [hunk 14]
```diff
@@ -55,50 +134,134 @@ export class CreateSnippetComponent {
...
+    this.repoName = this.activatedRoute.snapshot.params['repoName'];
+    this.repoOwner = this.activatedRoute.snapshot.params['repoOwner'];
+
+    if (pullNumber) {
+      this.isEditing = true;
+      this.getPullFileByPullNumber(this.repoName, this.repoOwner, pullNumber);
+    }
+
     this.filteredTags = this.snippetForm.get('tags').valueChanges.pipe(
       // tslint:disable-next-line:deprecation
```
Comment (NothingEverHappens) [inline apps/angular-thirty-seconds/src/app/create-snippet/create-snippet.component.ts]: why?

File: apps/angular-thirty-seconds/src/app/create-snippet/snippet-modal/snippet-overview.component.ts [hunk 15]
```diff
@@ -1,9 +1,24 @@
...
-import { MAT_DIALOG_DATA, MatDialogRef } from '@angular/material';
-import { SubscriptionLike } from 'rxjs/internal/types';
+import { MAT_DIALOG_DATA, MatDialogRef, MatSnackBar } from '@angular/material';
 import * as firebase from 'firebase';
+import { finalize, switchMap, takeUntil } from 'rxjs/operators';
+import { ReplaySubject } from 'rxjs/internal/ReplaySubject';
 import { SnippetService } from '../../shared/services/snippet.service';
+import { GitHubService } from '../../shared/services/github.service';
+
+interface SnippetInfo {
```
Comment (NothingEverHappens) [inline apps/angular-thirty-seconds/src/app/create-snippet/snippet-modal/snippet-overview.component.ts]: SnippetOverviewData

File: apps/angular-thirty-seconds/src/app/create-snippet/snippet-modal/snippet-overview.component.ts [hunk 16]
```diff
@@ -1,9 +1,24 @@
...
+  formValue: object;
```
Comment (NothingEverHappens) [inline apps/angular-thirty-seconds/src/app/create-snippet/snippet-modal/snippet-overview.component.ts]: Can we have proper type?

**Pull request merged**